### PR TITLE
python/darts: add new mentoring notes

### PR DESCRIPTION
**Pull request by IsaacG**
@BethanyG for a review, please.

### CODE REVIEW

File: tracks/python/exercises/darts/mentoring.md
Comment (kotp) [inline tracks/python/exercises/darts/mentoring.md:31]: This starts off with avoiding `else` and is specifically marked as code, but then talks about how it is temping to use the `elif` instead.  What does one have to do with the other?

They are not the same thing, and this is confusing at best.

File: tracks/python/exercises/darts/mentoring.md
Comment (kotp) [inline tracks/python/exercises/darts/mentoring.md:35]: While comute is misspelled, but also it may be misleading.  Is it supposed to be commute or compute?  Both meanings might be applied here, depending on what the exchange would result in.

File: tracks/python/exercises/darts/mentoring.md
Comment (kotp) [inline tracks/python/exercises/darts/mentoring.md:50]: Is it?  Clever, I mean.  I do not find it clever, really, but it surely is an alternative.  I do not even find it more complicated, just approached differently.

File: tracks/python/exercises/darts/mentoring.md
Comment (IsaacG) [reply tracks/python/exercises/darts/mentoring.md:31]: My bad. `s/else/elif`/

File: tracks/python/exercises/darts/mentoring.md
Comment (IsaacG) [reply tracks/python/exercises/darts/mentoring.md:35]: That was supposed to be `compute`.

File: tracks/python/exercises/darts/mentoring.md
Comment (IsaacG) [reply tracks/python/exercises/darts/mentoring.md:50]: Yeah. Agreed. It was something that didn't occur to me and it was surprising to me when I saw it. That triggered an emotional "that's clever" response. 

I'll go with the `git commit --amend; git push --force` route here.

File: tracks/python/exercises/darts/mentoring.md
Comment (IsaacG) [reply tracks/python/exercises/darts/mentoring.md:32]: > It might be worth talking about the difference in approach and why one might enjoy the disjointed expressions or the single complex expression.

You mean a nested ternary expression? I haven't seen this in the exercises I've mentored nor in the community solutions. While there's always other ways to solve exercises, that approach isn't something that comes up, probably because it's pretty messy and obviously so. It's not something that would be valuable for a mentor to bring up in this context, either, as it isn't helpful for this exercise.

Given that students don't use it and I don't think there's value in mentors raising it, I don't think it's helpful to mention in the notes.

> One might also wonder if there is a case equivalent

Python 3.10 introduces a case statement. I don't recall offhand if it supports comparisons. It's not super helpful for this exercise, though, as Exercism doesn't support 3.10 yet.

Comment (kotp) [reply tracks/python/exercises/darts/mentoring.md:32]: > You mean a nested ternary expression?

No, a single `if elif else end` statement.  Since you talk about avoiding it, might as well talk about the benefits of using it as well.

Comment (IsaacG) [reply tracks/python/exercises/darts/mentoring.md:32]: Given that the linter complains about it, I'm not sure that there's any reason to advocate for it for discuss it. Could you give an example of what would be useful to add to the mentoring guide on that topic for this exercise?

Comment (kotp) [reply tracks/python/exercises/darts/mentoring.md:32]: The linter also complains about the names `x` and `y` but for some reason it seems that is discounted.  It is apparently not the boss of us.

Without seeing the code that the linter complains about, I would be silly to rebuff that.  Instead, let me show code that the linter does not complain about using `elif`.

```python
import math

def score(x, y):
    distance = math.sqrt(x**2 + y**2)
    if distance <= 1:
        result = 10
    elif distance <= 5:
        result = 5
    elif distance <= 10:
        result = 1
    elif distance > 10:
        result = 0
    return result
```

It is single if statement, perhaps emulating a switch or case statement in other languages, and I do not find it is saying to avoid `elif`.

So, why are we saying to avoid `elif`?  Why are we not suggesting proper ways to use it, discussing if we should do the early returns, or if we should let the statement complete.

A return from a conditional branch theoretically does not allow the if statement to complete its work.  (Short-circuiting can be a good thing, so if that is a feature, we point that out.)

So, maybe the solution shown in this comment is truly "crap code".  Sometimes bad code is good information for mentoring.

Comment (IsaacG) [reply tracks/python/exercises/darts/mentoring.md:32]: Would it help if I was explicit about what I'm referring to? It's the above code but with the `if` replaced with `elif`:

```
def score(x, y):
    distance = math.sqrt(x * x + y * y)
    if distance <= 1:
        return 10
    elif distance <= 5:
        return 5
    elif distance <= 10:
        return 1
    return 0
````

The linter is not wrong and the warnings are valid. `elif` should not be used in this case.

> So, why are we saying to avoid elif? Why are we not suggesting proper ways to use it, discussing if we should do the early returns, or if we should let the statement complete.

This isn't saying "always avoid elif". It is saying, for this exercise, a talking point that is often relevant is this particular use of `elif` which should be avoided. If the student submits code that looks like the above snippet, this is something the mentor can point out and talk about.

This document also isn't attempting to show all the various ways `if` and `elif` can be used. It's specifically providing a talking point for this exercise targeting a specific pattern which is often used.

> A return from a conditional branch theoretically does not allow the if statement to complete its work. (Short-circuiting can be a good thing, so if that is a feature, we point that out.)

Pardon? Are you suggesting that `return` statements should not be used in an `if` branch?

I'm unclear what you would like to see with this note or what to do here.

Comment (kotp) [reply tracks/python/exercises/darts/mentoring.md:32]: Why are we targeting a specific pattern, and not several patterns that are used?
(This is a big part of what I am getting at.)

Given that the example code is the one that is often used, why not talk
about how the code does not read and explore that as well.  Why is it
better or more common for multiple statements rather than using a single
if statement with the structure provided.

I was also curious as to why we would return from inside of a statement,
which stops the statement from completing.  I mean as a general
practice.

I suppose I could ask the same thing, but in the opposite way… Are you
suggesting that return statements should be used in an if branch?  I
know that they can be used, and I lean more toward they probably should
not be, but I do acknowledge that they can be.

Also, there may be nothing to change here.

Comment (IsaacG) [reply tracks/python/exercises/darts/mentoring.md:32]: > Why are we targeting a specific pattern, and not several patterns that are used?

Because this is a specific pattern that comes up often in this exercise and is considered worth fixing.

> Given that the example code is the one that is often used, why not talk about how the code does not read and explore that as well. 

I don't understand what the second half of this sentence means or what you're trying to say/suggest.

> Why is it better or more common for multiple statements rather than using a single if statement with the structure provided.

I don't follow? Do you mean

```
if cond_a:
    return val_a
if cond_b:
    return val_b
# vs
if cond_a:
    return val_a
elif cond_b:
    return val_b
```

If you're trying to compare some other patterns, I'm not sure what you mean.

> I was also curious as to why we would return from inside of a statement, which stops the statement from completing. I mean as a general practice.

This pattern? It's quite common and there's no reason not to use it. Why would we not do this?

```
if cond:
    return val
```

I suppose I could ask the same thing, but in the opposite way… Are you suggesting that return statements should be used in an if branch? I know that they can be used, and I lean more toward they probably should not be, but I do acknowledge that they can be.

This pattern? It's quite common and there's no reason not to use it. Why would we not do this?

```
if cond:
    return val
```

> Also, there may be nothing to change here.

Personally, I don't think there's anything to change. I'm just waiting for approval so this PR can be merged.

Comment (kotp) [reply tracks/python/exercises/darts/mentoring.md:32]: Because we can let the if statement complete its work without short circuiting the statement.  It probably was not written to "usually never end", and so it may be an accepted (ab)use of the statement.  I mean, it feels that way.  But it might be the python way.

But if there is no change to be made, that is fine as well.  (Better to have the mentor notes come in for many mentors to review it, and have it evolve.

So as a maintainer, I am OK with bringing it in without change, knowing that if changes are needed they will come.

File: tracks/python/exercises/darts/mentoring.md
Comment (kotp) [inline tracks/python/exercises/darts/mentoring.md:84]: Since we have seen the idea "because the linter complains, it is not worth advocating" this code block should be changed for that same reason.

It is a "useless else on loop" as there is no break statement.

Again, the loop is being short circuited, which is a good thing. It is emulating, apparently, what would be a switch or case statement in other languages (where their case statement is not a "waterfall" case statement).  But there is no break that is expected when short-circuiting a loop, from what feedback I am getting from the linter.

Instead, the loop can be responsible for accumulating the score (emulating waterfall style case statement) and we can return the work in its own statement, which makes it clear the loop is responsible for the score, and the return (or any other work needing to be done) is done elsewhere.

And without the reported misuse of `else`.

**Review by kotp — CHANGES_REQUESTED** [review]

File: tracks/python/exercises/darts/mentoring.md
Comment (IsaacG) [reply tracks/python/exercises/darts/mentoring.md:84]: `else` removed.

File: tracks/python/exercises/darts/mentoring.md
Comment (IsaacG) [reply tracks/python/exercises/darts/mentoring.md:32]: > Because we can let the if statement complete its work without short circuiting the statement. It probably was not written to "usually never end", and so it may be an accepted (ab)use of the statement. I mean, it feels that way. But it might be the python way.

We can. By introducing a variable and moving the return outside the loop. Or we could keep things simple.

This is a commonly used pattern and commonly accepted. That's why the linter complains about the `else` and not the `return`.

> But if there is no change to be made, that is fine as well. (Better to have the mentor notes come in for many mentors to review it, and have it evolve.
> So as a maintainer, I am OK with bringing it in without change, knowing that if changes are needed they will come.

Sounds good to me. I'll resolve this conversation then.

**Review by kotp — APPROVED** [review]

**Comment by IsaacG** [comment]
Can a maintainer kindly merge if no further changes are needed?
Thanks!

**Pull request merged**
